File: lib/src/main/java/vproxy/vserver/util/ByteArrayCharStream.java
package vproxy.vserver.util;

import vjson.CharStream;
import vproxy.base.util.ByteArray;

import java.nio.charset.Charset;

public class ByteArrayCharStream implements CharStream {
    private final char[] chars;
    private int idx = -1;

    public ByteArrayCharStream(ByteArray data, Charset charset) {
        this.chars = new String(data.toJavaArray(), charset).toCharArray();
    }

    @Override
    public boolean hasNext(int i) {
        return idx + i < chars.length;
    }

    @Override
    public char moveNextAndGet() {
        return chars[++idx];
    }

    @Override
    public char peekNext(int i) {
        return chars[idx + i];
    }
}

File: lib/src/main/kotlin/io/vproxy/lib/common/ByteArrayCharStream.kt
package io.vproxy.lib.common

import vjson.CharStream
import io.vproxy.base.util.ByteArray
import java.nio.charset.Charset

class ByteArrayCharStream(data: _root_ide_package_.io.vproxy.base.util.ByteArray, charset: Charset) : CharStream {
  private val chars: CharArray = String(data.toJavaArray(), charset).toCharArray()
  private var idx = -1
  override fun hasNext(i: Int): Boolean {
    return idx + i < chars.size
  }

  override fun moveNextAndGet(): Char {
    return chars[++idx]
  }

  override fun peekNext(i: Int): Char {
    return chars[idx + i]
  }
}
